feat: Support root nameserver and additional record types

maikelpoot opened the pull request:
This PR will add support for all additional record types transip supports at the moment:

- Root NS
- ALIAS
- DS
- NAPTR
- TLSA

maikelpoot commented:
Ideally the caveats would be catched/fixed in the planning/validating phase but couldn't find any documentation or examples how to.

ross commented:
> Ideally the caveats would be catched/fixed in the planning/validating phase but couldn't find any documentation or examples how to.

`_process_desired_zone` is the place to do that sort of checking, it's requirements are in the docstring:

https://github.com/octodns/octodns/blob/9f73103300927733c691bf765866ec5c6f2e5784/octodns/provider/base.py#L35-L52

You can search the org for what various providers are doing/checking in that function:

https://github.com/search?q=org%3Aoctodns%20_process_desired_zone&type=code

The azure provider looks like it has something that's fairly close to the root caveats:

https://github.com/octodns/octodns-azure/blob/374f95986d884dc603ba0dbe55226c2b6595cf6d/octodns_azure/__init__.py#L1343

As for the TTL bit I'm not sure what makes sense. It could always warn if there's a non-matching value and just go ahead with that value or it could do the standard `supports_warn_or_except` behavior.

maikelpoot commented:
`_process_desired_zone` is indeed what i was searching for. 
Refactored the code to validate the root nameserver in `_process_desired_zone`

maikelpoot commented:
Changed it to draft for now, The opt-in for NS is a pain in .... when dealing with larger sets of domains in the config.

Trying to refactor to something like the googlecloud provider:
```
WARNING GoogleCloudProvider[google_public] root NS record supported, but no record is configured for example.com.
 INFO  GoogleCloudProvider[google_public] root NS record in existing, but not supported or not configured; ignoring it
 ```

ross commented:
> Trying to refactor to something like the googlecloud provider:

IIRC that's just using the default/built-in root handling of the base class. 

https://github.com/octodns/octodns/blob/9f73103300927733c691bf765866ec5c6f2e5784/octodns/provider/base.py#L152-L166

maikelpoot commented:
@ross Thanks for the info. 
Removed the opt-in code, updated all the tests and tested the code with some real-world config.
So i think this one is ready for review.

ross commented on octodns_transip/__init__.py:
```diff
@@ -116,12 +201,40 @@ def populate(self, zone, target=False, lenient=False):
+            if record._type == 'NS' and record.name == '':
+
+                # Check values for FQDN, IP's are not supported
+                values = record.values if record.values else [record.value]
```
I believe `record.values` should be sufficent here since the `_type` is being checked above and `NS` records should always have the values property.

```suggestion
                values = record.values
```

ross commented on octodns_transip/__init__.py:
```diff
@@ -116,12 +201,40 @@ def populate(self, zone, target=False, lenient=False):
+
+                # Check if TTL differs
+                if record.ttl != self.ROOT_NS_TTL:
+                    record.ttl = self.ROOT_NS_TTL
```
Point 3 in the `_process_desired_zone` guidelines calls out that the records should not be modified. 

https://github.com/octodns/octodns/blob/a1456cb1fcf00916ca06b204755834210a3ea9cf/octodns/provider/base.py#L35-L52

You can look around for examples, but generally you'll want to `record = record.copy()` and then modify the copy that's returned.

https://github.com/octodns/octodns/blob/a1456cb1fcf00916ca06b204755834210a3ea9cf/octodns/record/base.py#L273-L284

maikelpoot replied:
Applied the record.copy function

ross commented on octodns_transip/__init__.py:
```diff
@@ -135,11 +248,44 @@ def _apply(self, plan):
+            record = change.data['new']
+
+            if change.data['name'] == '' and change.data['record_type'] == 'NS':
+                values = (
+                    record.get('values')
+                    if record.get('values')
+                    else [record.get('value')]
+                )
```
Same here, since it's an NS should be able to just do `.values`.

```suggestion
                values = record.values
```

maikelpoot replied:
Applied and reverted, this broke the tests.
Tried `values = record.get('values')` but that broke when i altered the unit.tests.yaml to have a single value NS record
```
? ''
:  - type: NS
    value: ns0.transip.net.
```

```
                values = record.get('values')

                nameservers = []
>               for value in values:
E               TypeError: 'NoneType' object is not iterable

octodns_transip/__init__.py:268: TypeError
```

ross replied:
Ah. I didn't notice that a couple of lines above, ~257 that it's serializing the change using the `.data` property.

```
            record = change.data['new']
```

That's what's causing the problem/complications. That turns everything into a `dict` which loses the all of the `Record` object's behavior.  

If you apply the above change. This one should work correctly. (I've tested it locally now.)

ross commented on octodns_transip/__init__.py:
```diff
@@ -31,16 +33,48 @@ class TransipNewZoneException(TransipException):
     MIN_TTL = 120
     TIMEOUT = 15
     ROOT_RECORD = '@'
+    ROOT_NS_TTL = 3600  # Bogus value
```
Maybe better to comment something like "See root NS handling in _process_desired_zone" for more information. Just seeing "Bogus value" here wouldn't really make sense to people who don't already know that the TTLs aren't configurable.

maikelpoot replied:
Updated the comments to better inform the usage of the variable

ross reviewed (commented):
Overall looks good. Few minor things inline and it should be ready to 🚢 

Thanks.

ross commented on octodns_transip/__init__.py:
```diff
@@ -135,11 +253,44 @@ def _apply(self, plan):
             )
 
+        for change in changes:
+            record = change.data['new']
```
```suggestion
            record = change.new
```

maikelpoot commented:
Thought i tried that before without success, but maybe did something else wrong then. 

Removed the usage of .data for both the `record=` and the `if change.data['name'] == '' and change.data['record_type'] == 'NS':` using the record from "new" to check.

ross commented:
Thanks.

maikelpoot commented:
@ross Do you have an idea when the next release of the provider will be (and/or when it will be included within the docker containers, github action etc etc)

ross commented:
/cc https://github.com/octodns/octodns-transip/pull/42
